### PR TITLE
Revert #13862: Fail at link time if exception catching is e…

### PR DESCRIPTION
**Pull request by sbc100**
…nabled at compile time but not link time

In #13862 we stopped including exception catching code when
linking with `DISABLE_EXCEPTION_CATCHING`.  Otherwise closure
would complain about undefined references to `__cxa_is_pointer_type`
and `___cxa_can_catch`.

Instead, this patch unconditionally includes those two function
in `libc++abi` even in the `noexcept` version of that library.
Since these are normal native symbols they should only be linked
in if needed so there should be no overhead for application that
don't need them.

Since these symbols are unconditional dependencies of the exception
handling JS functions, but `libc++abi` is not always linked (for example
if `-nostdlib++` is passed) I also needed to make library_exceptions.js
conditionally included only if we are also linking `libc++abi`.   Same
with the unwind JS functions.  This allows `INCLUDE_FULL_LIBRARY`
to work with `-nostdlib++`.

### CODE REVIEW

**Comment by sbc100** [comment]
> …nabled at compile time but not link time
> 
> In #13862 we stopped including exception catching code when
> linking with DISABLE_EXCEPTION_CATCHING. Otherwise closure
> would complain about undefined references to `__cxa_is_pointer_type`
> and `___cxa_can_catch`.
> 
> Instead this patch unconditionally includes those two function
> in libc++abi even in the noexcept version of that library.
> Since these are normal native symbols they should only be linked
> in if needed so there should be no overhead for application that
> don't need them.

File: emcc.py [hunk 1]
```diff
@@ -2185,9 +2185,10 @@ def check_memory_setting(setting):
 
   if not settings.DISABLE_EXCEPTION_CATCHING:
     settings.EXPORTED_FUNCTIONS += [
-      # If not for LTO builds, we could handle these by adding deps_info.py
-      # entries for __cxa_find_matching_catch_* functions.  However, under
-      # LTO these symbols don't exist prior the linking.
+      # For normal builds that entries in deps_info.py are enough to include
```
Comment (kripken) [inline emcc.py:2188]:
```suggestion
      # For normal builds the entries in deps_info.py are enough to include
```

File: emcc.py [hunk 2]
```diff
@@ -2185,9 +2185,10 @@ def check_memory_setting(setting):
...
+      # these symbols whenever __cxa_find_matching_catch_* functions are
+      # found.  However, under LTO these symbols don't exist prior the linking
```
Comment (kripken) [inline emcc.py:2190]:
```suggestion
      # found.  However, under LTO these symbols don't exist prior to linking
```

File: emcc.py [hunk 3]
```diff
@@ -2185,9 +2185,10 @@ def check_memory_setting(setting):
...
+      # do we include then unconditionally when exceptions are enabled.
```
Comment (kripken) [inline emcc.py:2191]:
```suggestion
      # and we include then unconditionally when exceptions are enabled.
```

File: src/library.js [hunk 4]
```diff
@@ -3351,6 +3351,7 @@ LibraryManager.library = {
     return Number(str);
   },
 
+#if LINK_AS_CXX
```
Comment (kripken) [inline src/library.js:3354]: this seems surprising as libunwind is not a C++ library, it's a C API. Is it because only C++ uses it? (are we sure nothing else does?)

File: system/lib/libcxxabi/src/private_typeinfo.cpp [hunk 5]
```diff
@@ -1290,7 +1290,7 @@ __base_class_type_info::search_below_dst(__dynamic_cast_info* info,
 
 // XXX EMSCRIPTEN
 
-#ifdef __USING_EMSCRIPTEN_EXCEPTIONS__
+#ifndef __USING_WASM_EXCEPTIONS__
```
Comment (kripken) [inline system/lib/libcxxabi/src/private_typeinfo.cpp:1293]: when are the functions below used when exceptions are off? maybe add a comment.

File: tools/deps_info.py [hunk 6]
```diff
@@ -198,8 +198,18 @@
...
+    _deps_info['__cxa_find_matching_catch'] = ['__cxa_can_catch']
+    _deps_info['__cxa_find_matching_catch_1'] = ['__cxa_can_catch']
+    _deps_info['__cxa_find_matching_catch_2'] = ['__cxa_can_catch']
+    _deps_info['__cxa_find_matching_catch_3'] = ['__cxa_can_catch']
+    _deps_info['__cxa_find_matching_catch_4'] = ['__cxa_can_catch']
+    _deps_info['__cxa_find_matching_catch_5'] = ['__cxa_can_catch']
+    _deps_info['__cxa_find_matching_catch_6'] = ['__cxa_can_catch']
+    _deps_info['__cxa_find_matching_catch_7'] = ['__cxa_can_catch']
+    _deps_info['__cxa_find_matching_catch_8'] = ['__cxa_can_catch']
+    _deps_info['__cxa_find_matching_catch_9'] = ['__cxa_can_catch']
```
Comment (kripken) [inline tools/deps_info.py:212]: can write a loop here? but, why is this needed in this PR?

File: system/lib/libcxxabi/src/private_typeinfo.cpp (hunk 5 above)
Comment (sbc100) [reply system/lib/libcxxabi/src/private_typeinfo.cpp:1293]: When some object files were compiled with exception catching enabled.. but we do not enable exception catching at a link time.   Basically exactly the thing that this PR is allowing.   I agree a comment would be good.

File: tools/deps_info.py (hunk 6 above)
Comment (sbc100) [reply tools/deps_info.py:212]: When exception catching is enabled we force the inclusion of these symbols in emcc.py.   However when exception catchi is disabled at link time we still want to support the used of these functions become some object might have been builld with exception catching enabled.   

Rather then just always include these functions I decided to go back to old way to include these dependencies prior to #13528.   This method works for the non-LTO builds.. LTO build will be handled by the forces dependency `emcc.py`.

File: src/library.js (hunk 4 above)
Comment (sbc100) [reply src/library.js:3354]: No but it is a C++ support library that is excluded from the link when you build with `-nostdlib++`.   Whenever we exclude the native part we also need to exclude the JS part.

**Comment by sbc100** [comment]
Maybe I can/should split out of the LINK_AS_CXX stuff .. it seems separable enough.

**Review by kripken — APPROVED** [review]
Thanks for the explanations, makes sense to me now.
